### ISSUE TITLE
CompoundDocument leaks a file descriptor

### ISSUE THREAD

**Issue by tingley** [haraldk/TwelveMonkeys#438]
The `CompoundDocument` class uses a `LittleEndianRandomAccessFile` internally to read the file contents:
https://github.com/haraldk/TwelveMonkeys/blob/24c6682236e5a02151359486aa4075ddc5ab1534/common/common-io/src/main/java/com/twelvemonkeys/io/ole2/CompoundDocument.java#L105

However, `input.close()` is never called, which means that the underlying `RandomAccessFile` can hold an open file descriptor indefinitely in a long-running process.

Is there any reason not to just add a `CompoundDocument#close()` method?

**Comment by haraldk** [comment]
Hi Chase,

Thanks for reporting!

I think you are right. I wrote this code ages ago when I had a habit of adding convenience methods for everything... Which, of course, turned out to be less than ideal. None of my own code actually uses the constructor that takes a `File` as parameter at the moment, except for the unit tests for the `CompoundDocument` class itself. So maybe it should just be deprecated for now and deleted for the 4.0 version. Or do you use it for anything?

Adding a `close()` method and perhaps implementing `AutoClosable` is also possible.

But what should the method do in cases where it didn't open the stream/file handle? No-op? Or close the stream anyway? I'm not sure, and I think the API will be kind of confusing either way... Which is why I think removing the method would be the best option (perhaps exposing a new constructor for `LittleEndianRandomAccessFile` should anyone really, really need it) and handle all stream/file closing issues externally to the class.


Best regards,

-- 
Harald K

**Comment by tingley** [comment]
Hi Harald,

I think I reported this wrong.  We use `CompoundDocument` in a weird way, basically to detect if an OpenXML file is encrypted or not.  You can see the code here:

https://bitbucket.org/okapiframework/okapi/src/41887d36930ebb5ecb2b6f62ba1d5cd2e279b2f3/okapi/filters/openxml/src/main/java/net/sf/okapi/filters/openxml/OpenXMLFilter.java#lines-452

So we're using the constructor that takes an `InputStream` after all.  But this then uses a `FileCacheSeekableStream` that creates a temp file and then uses a `FileCache` to access it.  If I'm following this correctly, it looks like `FileCache` has the same instance with never closing its internal `RandomAccessFile` instance.

**Comment by haraldk** [comment]
Thanks for following up on this!

That does indeed seem to be the case. Which is even worse...

My own code (the `ThumbsDBImageReader`) only uses the constructor taking an `ImageInputStream` parameter. And I believe this one is safe, as long as you close the `ImageInputStream` instance (which you can do externally). 

Something like: 

````java
    private boolean isZipFileEncrypted(File file) throws IOException {
        // If you pass the File parameter to the CompoundDocument constructor it
        // opens a file pointer that is not released properly and remains open for
        // the entire lifetime of the application. Using an FileImageInputStream we take
        // control of the file pointer and release it properly after the check.
        try (ImageInputStream is = new FileImageInputStream(file)) {
            new CompoundDocument(is);
            return true;
        } catch (CorruptDocumentException e) {
            return false;
        }
    }
````

This could perhaps work as a workaround for you, until I get the issue fixed. 


Best regards,

-- 
Harald K

**Comment by tingley** [comment]
Thanks, that's extremely helpful!  I'll give it a try.

**Comment by haraldk** [comment]
Hmmm...

The `CorruptDocumentException` doesn't really handle the byte order mark correctly, it just assumes that the byte order is little endian (Intel). And the `ImageInputStream` probably has big endian byte order.
Try adding:

````java
            is.setByteOrder(ByteOrder.LITTLE_ENDIAN);
````
...before creating the `CompoundDocument`.  I think that should work.

All the other constructors will create streams with the byte order explicitly set to little endian (in addition to leaking file descriptors...). Sorry for making such a mess... :-P


-- 
Harald K

**Comment by tingley** [comment]
That seems to work.  Tremendous, thanks for all your help!

**Comment by haraldk** [comment]
Excellent! 

I'll keep this issue open, to track the issues you discovered.

A quick summary:

- [ ]  `new CompoundDocument(File)` leaks file descriptors, as the `RandomAccessFile` is never closed
- [ ]  `new CompoundDocument(InputStream)` leaks file descriptors, as the `FileCacheSeekableStream` never closes its `FileCache` (`RandomAccessFile` backed)
- [ ]  `new CompoundDocument(ImageInputStream)` does not set the byte order to `LITTLE_ENDIAN`

The first two can't be worked around in user code, and must be fixed somehow. Perhaps by using in-memory caching, or by adding a `close()` method.

The last is easily worked around, by explicitly setting the byte order prior to creating the compound document.

-- 
Harald K

**Comment by haraldk** [comment]
Fix for this is in 3.4-bugfix branch and master.
